### ISSUE TITLE
IRReturnJump:<null:> on return inside instance_eval

### ISSUE THREAD

**Issue by AlBundy33** [jruby/jruby#5349]
I'm using jruby-dist-9.1.17.0-bin.zip because I need Java 7 compatibility.

For some testing I run scripts with instance-eval because I need to access the variables from my "main"-context which works fine.

But if I do a return in my "evaled" script, I get an exception: org.jruby.embed.EvalFailedException: IRReturnJump:<null:>

Here is a simple testcase:
```
$msg = "Hello World"
script = "puts $msg\nreturn"
instance_eval(script, "test")
```
And this is the jruby-part of the stacktrace:
```
        IRReturnJump:<null:>
        org.jruby.embed.internal.EmbedEvalUnitImpl.run(EmbedEvalUnitImpl.java:137)
        org.jruby.embed.jsr223.JRubyEngine.eval(JRubyEngine.java:118)
        org.jruby.embed.jsr223.JRubyEngine.eval(JRubyEngine.java:138)
```

**Comment by headius** [comment]
Given that we've had no maintenance releases of JRuby 9.2 yet we might consider doing a 9.1.18.0. I  believe originally we planned to, but then got too busy on 9.2 work over the summer.

However, this might not be an easy fix. In working on 9.2, various aspects of non-local flow control (like this return) were redesigned, and 9.2.0 does not exhibit this problem.

A workaround for you would be to put the body of the eval in a method and call the method, or do the same with a block:

```ruby
$msg = "Hello World"
script = "->{ puts $msg\nreturn }.call"
instance_eval(script, "test")
```

Why do you still need Java 7?

**Comment by AlBundy33** [comment]
Using the block seems to be a suitable workaround.

Unfortunately we're sticking with Java 7 (at least on client-side) due to our customers.
So we're trying to be as long as possible Java 7 compatible.

**Comment by headius** [comment]
@AlBundy33 I feel your pain, sorry we can't help  on the Java 7 front. 

Given that this appears to be fixed in current supported versions of JRuby, I'm going to mark this as fixed in 9.2. Up to @enebo if we should try to investigate this for a 9.1.18 release.

**Comment by AlBundy33** [comment]
@headius: Using the workaround is ok for me but for completness: This is an issue in 9.2.0.0 too.

```
irb(main):001:0> JRUBY_VERSION
=> "9.2.0.0"
irb(main):002:0> $msg = "Hello World"
=> "Hello World"
irb(main):003:0> script = "puts $msg\nreturn"
=> "puts $msg\nreturn"
irb(main):004:0> instance_eval(script, "test")
Hello World
Unhandled Java exception: IRReturnJump:<Static Type[1898220577]: local []:>
```
